### DISCUSSION TITLE
Reputation score system for Bisq Satoshi Square app

### DISCUSSION THREAD

**Discussion by chimp1984** [bisq-network/bisq2#201] [General]
We should find an easy to understand signalling mechanism for the trustworthiness of traders in the chat based trade system in the Satoshi Square app.

Initially we likely use burned BSQ as main source for reputation. But we could add more over time like:
- Account age from accounts the user had on Bisq 1
- Signed account age witness from accounts the user had signed in Bisq 1
- BSQ bonds

Each category/source of reputation will have a different level of security. E.g. Signed account age witness can be considered more powerful than Account age. Burned BSQ might be considered more powerful than BSQ bonds (at same amounts).

So we should avoid a single term for traders who has provided one or more sources of reputation (like "trusted seller" or "premium trader").
We also should avoid hierarchical categories assigned to the categories like gold, silver, bronze as it would cause problems once we want to add a more secure source (ok plantinum can top gold, but what comes after that?).

The relevant numbers should be normalized by using a relative score. E.g. if there are 10 traders with burned BSQ ranging from 100 BSQ to 1000 BSQ, the one with 1000 BSQ can be considered 10x more trustworhty that the one with 100 BSQ, thus we could normalize those numbers to a scale of 1-10.

We should consolidate multiple categories to one single score.
E.g. How to compare trustworthiness of a trader with 100 BSQ burned to one who has 2 accounts signed with account age of 100 days and 300 days?
We could set a weight per category (say burned BSQ has weight 10 and signed account age witness category has weight 15) and then multiply the relevant numbers to get the total score:
E.g.
- 100 BSQ burned = 100 x 10 = 1000 total score
- 100 days x 15 + 300 days x 15 = 1500 + 4500 = 6000 total score

The newbie trader do not need to understand and estimate how much trustworthiness each category carries (its a hard task anyway). But just take the total score number which again can be normalized to a scale of 1-10 and be caclulated using the relative / temorary situation. E.g. if there are many trustworthy traders online competition to get a high score might be higher than in times when less trustworthy traders are online. So ones trust score floats when peers are joining and leaving (maybe we need to avoid to show frequent updates to the users to not confuse them).

With all that inherent complexity the main goal need to be that its very simple to understand for the newie user.
So to have attached a single score (e.g. 1-10 scale) should fulfill that requirement. How that is calculated is not that important for newbies. 

For pro traders who want to earn reputation, they have to learn a bit more about the options and the pro/cons (e.g. privacy is weaker if they connect to Bisq 1 accounts). But this is actually a kind of protection to get only experienced Bisq traders into that group as it is important for the human onboarding aspect that they understand well the concepts used in Bisq.

We could adopt the trader type screen:
- I am new to Bisq/Bitcoin. -> I want to make my first BUY offer.
- I am an experience Bisq trader. -> Learn more how to build up reputation

Maybe we could find a term which represent that score more as a reputation token. In fact its is a kind of token which gets defined in a decentalized and verifyable way and is composed of multiple inputs. One can mine such a token using various mechanisms like burning/bonding BSQ or being an active trader and utilizing the supported security features. 

Any suggestions for a snappy name for such a token (score)? 
"RepuTok", "Bisquitte", "Bisquette"?

**Comment by alkum** [comment]
How about ETP - Earned Trust Points?

**Reply by chimp1984** [reply]
I would prefer to avoid that users need to learn some new term and `Earned Trust Points` is too long for display in some areas. Also not so sure if we should use `trust` as it might be too much of interpretation we insert by that.

**Comment by chimp1984** [comment]
## Design specification for the reputation domain

There are multiple reputation sources which carry different `weight`s.
Each `ReputationSource` contains some `Proof` for verification.
The total reputation score is the sum of all reputation sources using their values multiplied with the weight.


### Burned BSQ

We apply a time based decay function to the burned BSQ amount by declining over 30 days to 0.

`Weight` : 1000 
`ReputationSource`: `BURNED_BSQ`

Example:
A single 100 BSQ burn would have following score values over time:
At burn time: 100 * 1000 = 100 000
After 30 days: 0

Proof is provided by a BSQ explorer lookup for the TX ID and the identity pubKeyHash which was used as preimage for the hash. The hash is in the opReturn data of the tx. Lookup is only done once by user and locally persisted afterwards.
Verification is only done at interaction time (e.g. opening a private chat, entering a trade).

### Burned BSQ age

We use the age and value of the burned BSQ as a monotonily growing reputation score. The burn BSQ transaction serves as timestamp.

The score is calculated simply by multipying age as number of days with amount of burned BSQ and multiplied by the weight.

`Weight` : 1
`ReputationSource`: `BURNED_BSQ_AGE`

Example:
A single 100 BSQ burn would have following score values over time:
At burn time:  0
After 30 days: 100 * 30 * 1 = 3 000
After 120 days: 100 * 120 * 1 = 12 000
After 360 days: 100 * 360 * 1 = 36 000


### Identity age

Once we have an Open Time stamp service integrated we can timestamp the date when an identity was created (similar to account age in Bisq 1). We can use that as a monotonily growing reputation score.

The score is calculated simply by age as number of days multiplied by the weight.

`Weight` : 100 
`ReputationSource`: `IDENTITY_AGE`

Example:
A new identity would have following score values over time:
At identity creation time: 0
After 30 days: 30 * 100  = 3000
After 60 days: 60 * 100 = 6000
After 120 days: 120 * 100 = 12 000
After 360 days: 360 * 100 = 36 000

### Imported Bisq 1 Account age

Once we have a tool for importing and notarizing Account age data from Bisq 1 we can apply similar to Identity age the Bisq 1 account age to be used for the globla reputation score. This will come with costs on privacy as the user would link their past Bisq 1 identity to the new identiy. Not sure if that feature is wortht to implement, but lets add it as potential candidate. 
Growth function same as for Identity age.
Global weight for Identity age: 50 (we keep it lower to not create too much incentive for a privacy degrading behaviour)

`Weight` : 50 
`ReputationSource`: `BISQ1_ACCOUNT_AGE`

### Imported Bisq 1 Signed account age witness

Once we have a tool for importing and notarizing Signed account age witness data from Bisq 1 we can use that for the global reputation score. Similar like account age it comes with some costs on privacy, but here it serves a more relevant security feature which is substantial for the 2of2 Multisig protocol. So it might be that once we have implemented the 2of2 Multisig protocol that we need that concept anyway (or maybe not as the risk for stolen bank account seems to have diminished to the obligatory use of 2FA for banks). When implementing it it needs to be reviewed if there are some improvements possible to reduce privacy issues.

`Weight` : TBD 
`ReputationSource`: `BISQ1_SIGNED_ACCOUNT_AGE_WITNESS`


## Total score

The total reputation score would be the sum of all sources. 

Example:
A user creates their identity and burns each 30 days 100 BSQ.
Total reputation score is sum of burned BSQ score + accumulated Burn age + identity age:
At burn time: 100 * 1000 + 0 + 0= 100 000
After 30 days: 100 * 1000 + 100 * 30 + 30 * 100 = 106 000 (they burned again 100 BSQ)
After 60 days: 100 * 1000 + 100 * 60  + 100 * 30 + 60 * 100 = 115 000
After 90 days: 100 * 1000 + 100 * 90  + 100 * 60  + 100 * 30 + 90 * 100 = 127 000
After 240 days: 100 * 1000 + 30*100*8*(8+1)/2   + 240 * 100 = 100 000 + 108 000 + 24 000 = 232 000
After 360 days: 100 * 1000 + 30*100*12*(12+1)/2   + 360 * 100 = 100 000 + 234 000 + 36 000 = 370 000

After 8 Months the accumulated age factor is getting larger than the factor from the newly burned BSQ.

The function for the accumulated `BURNED_BSQ_AGE` is: n * (n+1)/2 where n is multiple of 30 days (roughly a month)

A user who have burned only once 100 BSQ and then never again would have after 360 days:
After 360 days:  100 * 360 + 360 * 100  = 72 000

Not sure if the influence of accumulated score is not too strong? 
Beside the core function to serve as security, we want to incentivice burning BSQ as it serves as fee payment machanism.
Statically seen one could argue that after several month there is not much incentive anymore to burn BSQ as the accumulated score is already good, but one has to consider the competition between sellers. A long term seller who burned each month 100 BSQ would be after about 8 months without burning new BSQ on the same score level than a new user who just burned 100 BSQ. 

I guess it will be hard to find the right values here without real life experience/feedback.
So we need to keep the system flexible to be able to update the factors and maybe even the algorithm without breaking consensus.
The facors could be deployed to the network as authroized P2P data (similar like filter data in Bisq1). Changing the algorithm would be more difficult and would require some versioning concept.

**Comment by pazza83** [comment]
Hi @chimp1984 thanks for documenting this.

I have some questions so I can better understand how the new mechanism will work.

- How do you propose the sum of all multiple weighted scores would be normalized to a scale of 1-10? Would the 1-10 score be based on the scores of other users (eg work out users' percentile relative to each other and convert that to an integer between 1 to 10?)

- One of the aspects that I like about Bisq 2 is the ability to create multiple identities. Looking at the reputation score system managing multiple identities might come with significant costs. What if any weighted values would be shared between identities? 

- Similar to the above for the Imported Bisq 1 Account age and Imported Bisq 1 Signed account age witness users will have multiple accounts ages / signed witnesses how would these be applied to their various identities? 

- Would it be 1 imported account per identity or could it be multiple?

- The investment a user will make into their identity could be very significant. Reading the above makes me think the reputation score system would be suitable for use with the same limits per payment method as Bisq 1. At least proportional to their 1-10 score. Is this the intention?

- Will poor behavior of users (eg late release of BTC, or late payment) lead to penalties to reduce a user's reputation score?

- Can makers choose to have their offers taken by users with a certain reputation. Eg I want to sell users with a reputation of 5 or more 0.01 BTC at 5% over market rate. But for users with a reputation of less than 5, I want to give them a markup of 12% to cover the additional risk of them not paying, leaving a Bitcoin payment reference etc

**Comment by chimp1984** [comment]
> How do you propose the sum of all multiple weighted scores would be normalized to a scale of 1-10? Would the 1-10 score be based on the scores of other users (eg work out users' percentile relative to each other and convert that to an integer between 1 to 10?)

Ah good point, did not cover that yet. Yes basic idea was to check available users reputation and rank them among them. But have not thought it out so far. Maybe it will not work that easy as it might not be that clear to detect which users have available offers. Maybe all users of all trade chat messages are used for the relative ranking. 

> One of the aspects that I like about Bisq 2 is the ability to create multiple identities. Looking at the reputation score system managing multiple identities might come with significant costs. What if any weighted values would be shared between identities?

I fear there is no way to get both. Reputation is tied to identity and mixing/connecting identities will merge them to one parent identity. Maybe with some zero knowledge magic it will be possible at some point but thats beyond me skill set.

> Similar to the above for the Imported Bisq 1 Account age and Imported Bisq 1 Signed account age witness users will have multiple accounts ages / signed witnesses how would these be applied to their various identities?

I have not worken on the details but if one would import those identity data into multiple Bisq2 identities they would merge/connect those identities, thus eliminating the benefits of separating them in the first place. Sure for superficial views from other users it might still serve some purpose but if one wants to find out they can if there is some connection.

> Would it be 1 imported account per identity or could it be multiple?

As said I have not worked on the details in that area. But I assume we can be flexible and allow to import multiple Bisq1 identities and also allow to import them to multiple Bisq2 identities (though would eliminate the benefit of having multiple identities).

> The investment a user will make into their identity could be very significant. Reading the above makes me think the reputation score system would be suitable for use with the same limits per payment method as Bisq 1. At least proportional to their 1-10 score. Is this the intention?

Thats  an interesting thought, I have not got that far but maybe it makes sense to tie the limits to the sellers score. But it will be hard to find out what are safe values. At the end its still weak security and vulnerable to a long con attack (user build up good score and then hit with a parallelized scam - only expected low liquidity will serve as a limiting factor here).
I think its more prudent to start cautious and expand it when we get more experience. Also the high premium price will limit the usefulness as well and will move users to more secure protocols.
I guess its better that we stay focussed on the main goal - easy onboarding for newbies for small amounts. 
But we can reevaluate that after we see how well it works in real life.

For the limit: I have not thought much about it but basic idea was to make a hard coded limit (about 200 EUR) and apply that to the offer creation page. We cannot prevent that users write in the chat higher limits or that they trade higher limits in the private chat. But it can be clearsly stated that its not recommended and the system is not designed for high value trades.  
To connect it to reputation scores might add too much complexity - users will not understand how the limit is calculated.

> Will poor behavior of users (eg late release of BTC, or late payment) lead to penalties to reduce a user's reputation score?

Hm...I fear that would introduce too much centralized control (from mediators or whoever has the power to influence that).
I think its better to limit that to clear scams. For clear proofs of a scam the mediator can request that this users identity get banned thus their reputation is useless as it is tied to the identity. This ban will work similar like in Bisq1 with the filter, thus is a bit centralized (though it can be a bonded role and there could be multiple).

Users are free to share any bad experience. Giving positive feedback comes with privacy issues (one can deduct how many trades a user had if thats public data), so I perfer to not introduce that, beside pure private feedback back to the trade peer.

There will be a bit of a challenge where the boundaries of privacy protection and permission to provide feedback lies. I think its one of the open areas where we need to see how well it works in practice and adopt if needed. From what I hears about the existing Keybase experience it seems there have been basically no such problems, but surely it will be related to scale. 
Whatever rules we apply, we cannot avoid that frustrated users will share their feedback outside of the app (reddit, forum, matrix,...), so I think its better to provide them with some option but still make it clear that te feedback should be made in a way to respect the peers privacy as much as it makes sense inside the context.

> Can makers choose to have their offers taken by users with a certain reputation. Eg I want to sell users with a reputation of 5 or more 0.01 BTC at 5% over market rate. But for users with a reputation of less than 5, I want to give them a markup of 12% to cover the additional risk of them not paying, leaving a Bitcoin payment reference etc

Ah good point, forget to mention that.
Yes its planned that by default a newbie user has set a min. required reputation score which can be only achieved by burning some non trivial amount of BSQ (e.g. 50 BSQ). I would prefer simlicity here as well and not add more complex conditions. The user can lift or adjust the min. reputation score in the settings. This will protect them also from getting private messages from users who do not have that reputation score level, thus avoiding the risk that a scammer with no reputation tries to get in touch with newbies to trick them.
When the user lowers that level they will get a warning with context info.

So far we have kept out price info from the create offer screen to keep it simple and to not confront the newbie too early with the need to understand the context why a premium is justified. Otherwise I think newbies would like to trade at market rate as they first don't understand why sellers have to earn something on the premium for the service they provide. That info can be communicated when a private trade chat start (e.g. a first system message giving background info and security advices...)

**Comment by chimp1984** [comment]
## Transfer DAO data via REST API from Bisq 1 to Bisq 2 network

I started to work on a REST API for Bisq 1 to provide DAO data which are needed for the reputation system (burned BSQ).
Currently we provide only whats needed but it can grow to a full DAO node which can provide later when more trade protocols are using DAO/ BSQ features the required DAO data for Bisq 2.
There is a client app which requests those data (currently polling but could be long polling for push data later) and publish the data to the Bisq 2 P2P network as authorized data.

### Authorized data
Authorized data are network data which are signed by a private key where the public key(s) are hard coded in the application and nodes can verify that the data was signed by that authorized node who got provided the private key by the developers. It is similar as it is used in Bisq 1 for filter data, alerts (update notifications) or arbitrator/mediator registrations. It could be improved to use a bonded role and the role owner provides the public key themself as a PR to the source code so that even the developers do not have access to the private key. There can be multiple key holders and we could add a requirement for multiple signers depending on the type of data, mimicing the functionality of multisig (e.g. user could define how many signatures are required for accepting the data).

Authorized data have some trust issues but those are also in place for the alternative options like requesting the DAO data from the BSQ explorers (the explorer operator need to be trusted in that case). If a user wants to run a full DAO node they can eliminate those issues. 
Atm there is no implementation for such a full node setup but there are various options how to do it (e.g. allow user to pass by program argument a pubKey which is their own pubKey making themself an authorized data provider).

Also there will be an option to bypass those authorized nodes or ban certain pubkeys to protect against malicious actors.

### Burn BSQ transaction data
We utilize the amount and the date of the burn BSQ transaction. Those are the base for 2 different reputation sources (BURNED_BSQ, BURNED_BSQ_AGE). To detect which data are from Bisq 2 reputation use cases and which are other burn BSQ use cases (e.g. Burningman) we need to provide also the `ChatUser` as distributed data to the network.
By having the ChatUser data available in the network we can filter the proof of burn transactions for opReturn data containing a ChatUser IDs to get a collection of Bisq 2 proof of burn transactions.
This collection is used for verification of peers reputation data and for calculating the relative score. Without the requirement for relative scores we could have used a diferent model where we provide the data for verification only at the moment when a financial interaction starts (e.g. a private message is sent), thus not requiring the P2P data storage layer for the distribution of that data. But as we need a view over all active ChatUsers for calculating the relative score we need to distribute all of them. The ChatUser has a time-to-life which helps to limit the data load as well as filters out inactive users. 
The details about TTL and re-distributions are not decided yet, but goal is to have only active users in the collection.

This diverges a bit from the original goal to use only the trade competitors for the relative score ranking (e.g. all makers in a certain market). It still could be implemented that way but I guess the extra complexity and abuse risks with edge cases (markets with very few makers) are not worth it. So the current plan is to use all available ChatUser data as our pool for calculating the relative reputation of a user.  

Btw. the ChatUser is also required as we do not send the full ChatUser data in each public chat message (as it was in an earlier version) as that would create much larger message data and would be highly inefficient. By using only the ChatUser ID and distribute the ChatUser as separate data we avoid that inefficiency. The TTL of the Chatuser is tied  to the TTL of messages (and longer) to ensure that the ChatUser data is always longer available than the messages sent from that user.

Work on that is not completed yet but most of the basics are already in place.
